File: clients/typescript/src/polyfetch.ts
import { Worker } from "worker_threads";
import * as path from "path";
import * as fs from "fs";
import type { TargetRecord } from "./fetch";
import { writeReport } from "./report";

// Exit codes, per the contract (§9):
//   0 — run completed (per-target errors do NOT change this)
//   1 — the run itself failed to complete
//   2 — invalid invocation (bad flag, unreadable targets file)
const EXIT_RUN_FAIL = 1;
const EXIT_USAGE = 2;

interface Config {
  concurrency: number;
  targetsPath: string;
  baseUrl: string;
  timeoutMs: number;
}

function usageError(msg: string): never {
  process.stderr.write(`polyfetch: ${msg}\n`);
  process.exit(EXIT_USAGE);
}

function parseArgs(argv: string[]): Config {
  const cfg: Config = {
    concurrency: 8,
    targetsPath: "shared/targets.txt",
    baseUrl: "http://localhost:8080",
    timeoutMs: 30000,
  };

  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    const needValue = (name: string): string => {
      if (i + 1 >= argv.length) usageError(`missing value for ${name}`);
      return argv[++i];
    };
    switch (arg) {
      case "--concurrency": {
        const v = needValue("--concurrency");
        const n = Number(v);
        if (!Number.isInteger(n) || n < 1) usageError("--concurrency must be an integer >= 1");
        cfg.concurrency = n;
        break;
      }
      case "--targets":
        cfg.targetsPath = needValue("--targets");
        break;
      case "--base-url":
        cfg.baseUrl = needValue("--base-url");
        break;
      case "--timeout": {
        const v = needValue("--timeout");
        const n = Number(v);
        if (!Number.isInteger(n) || n < 0) usageError("--timeout must be an integer >= 0");
        cfg.timeoutMs = n;
        break;
      }
      case "-h":
      case "--help":
        process.stdout.write(
          "usage: polyfetch [--concurrency K] [--targets PATH] [--base-url URL] [--timeout MS]\n",
        );
        process.exit(0);
      default:
        usageError(`unknown flag: ${arg}`);
    }
  }

  return cfg;
}

// Read targets: one per line, skipping blank lines and lines beginning with
// '#'. The remainder of each line (including any "?size=...&delay=..." query) is
// preserved verbatim. Returns null if the file cannot be read.
function readTargets(p: string): string[] | null {
  let text: string;
  try {
    text = fs.readFileSync(p, "utf-8");
  } catch {
    return null;
  }
  const out: string[] = [];
  for (const raw of text.split("\n")) {
    const line = raw.trim();
    if (line === "" || line.startsWith("#")) continue;
    out.push(line);
  }
  return out;
}

/**
 * The concurrency core. Spawns a bounded pool of K worker threads and feeds each
 * one target at a time, so at most K targets are in flight — the contract's
 * concurrency bound. Each worker runs the full fetch + process pipeline on its
 * own thread (true CPU parallelism for the hashing stage).
 *
 * Contrast with the C++ client: there, the shared work queue needed a
 * std::mutex because multiple OS threads touched it. Here, the coordination
 * state (nextIndex, completed) lives on the single-threaded main event loop, so
 * message handlers run sequentially and NO lock is needed. The single-
 * threadedness that forces us into worker_threads for CPU work is, for the
 * coordination itself, a simplifying asset.
 */
function run(config: Config, targets: string[]): Promise<TargetRecord[]> {
  return new Promise((resolve, reject) => {
    const n = targets.length;
    const results: TargetRecord[] = new Array(n);
    if (n === 0) {
      resolve(results);
      return;
    }

    const k = Math.min(config.concurrency, n);
    let nextIndex = 0;
    let completed = 0;

    const workerPath = path.join(__dirname, "worker.js");
    const workers: Worker[] = [];

    const assignNext = (worker: Worker): void => {
      if (nextIndex < n) {
        const idx = nextIndex++;
        worker.postMessage({ index: idx, target: targets[idx] });
      }
      // else: no work left; the worker idles until terminated below.
    };

    for (let i = 0; i < k; i++) {
      const worker = new Worker(workerPath, {
        workerData: { baseUrl: config.baseUrl, timeoutMs: config.timeoutMs },
      });
      workers.push(worker);

      worker.on("message", (msg: { index: number; record: TargetRecord }) => {
        results[msg.index] = msg.record;
        completed++;
        if (completed === n) {
          // All targets processed; every worker is idle. Tear them down, then
          // resolve. We do not call process.exit here — letting the event loop
          // drain ensures the stdout write in writeReport flushes fully.
          Promise.all(workers.map((w) => w.terminate())).then(() => resolve(results));
        } else {
          assignNext(worker);
        }
      });

      worker.on("error", (err) => {
        reject(err);
      });
    }

    // Kick off: give each worker its first task.
    for (const w of workers) assignNext(w);
  });
}

async function main(): Promise<void> {
  const cfg = parseArgs(process.argv.slice(2));

  const targets = readTargets(cfg.targetsPath);
  if (targets === null) {
    process.stderr.write(`polyfetch: cannot read targets file: ${cfg.targetsPath}\n`);
    process.exit(EXIT_USAGE);
  }

  let records: TargetRecord[];
  try {
    records = await run(cfg, targets);
  } catch (err) {
    const e = err as { message?: string };
    process.stderr.write(`polyfetch: run failed: ${e.message ?? "error"}\n`);
    process.exit(EXIT_RUN_FAIL);
  }

  writeReport(records);
  // Natural exit (code 0) after stdout drains and workers are terminated.
}

void main();

File: clients/python/polyfetch.py
#!/usr/bin/env python3
"""polyfetch Python client.

Concurrency model: a pool of worker PROCESSES (multiprocessing via
ProcessPoolExecutor), each running the full fetch + process pipeline. Processes,
not threads, are what give Python true CPU parallelism: each process has its own
interpreter and its own GIL, so the hashing stage parallelizes across cores
without GIL contention. This is the structural change the experiment's thesis
predicts for Python — the direct analog of the TypeScript client reaching for
worker_threads to escape its single-threaded event loop. Go and C++ needed no
such change; one goroutine / OS thread handled both stages.

A thread-based variant (ThreadPoolExecutor) would be the key contrast: because
hashlib releases the GIL during large hashes, threads would parallelize the
SHA-256 work *partially*, while the pure-Python glue stays serialized — a more
nuanced picture than 'the GIL forbids CPU parallelism'. See process.py.
"""

import argparse
import sys
from concurrent.futures import ProcessPoolExecutor

from fetch import Record, init_worker, process_one
from report import write_report

# Exit codes, per the contract (§9):
#   0 — run completed (per-target errors do NOT change this)
#   1 — the run itself failed to complete
#   2 — invalid invocation (bad flag, unreadable targets file)
EXIT_RUN_FAIL = 1
EXIT_USAGE = 2


def parse_args(argv: list[str]) -> argparse.Namespace:
    # argparse exits with code 2 on its own for unknown flags, missing values,
    # and non-integer values for type=int — exactly the contract's usage-error
    # behavior. parser.error() (used for our extra checks) also exits 2.
    p = argparse.ArgumentParser(prog="polyfetch", add_help=True)
    p.add_argument("--concurrency", type=int, default=8,
                   help="max targets in flight simultaneously (>=1)")
    p.add_argument("--targets", default="shared/targets.txt",
                   help="path to targets file")
    p.add_argument("--base-url", dest="base_url", default="http://localhost:8080",
                   help="server base URL")
    p.add_argument("--timeout", dest="timeout_ms", type=int, default=30000,
                   help="per-target timeout in milliseconds")
    args = p.parse_args(argv)

    if args.concurrency < 1:
        p.error("--concurrency must be >= 1")
    if args.timeout_ms < 0:
        p.error("--timeout must be >= 0")
    return args


def read_targets(path: str) -> list[str] | None:
    """One target per line, skipping blank lines and lines beginning with '#'.
    The remainder of each line (including any '?size=...&delay=...') is preserved
    verbatim. Returns None if the file cannot be read."""
    try:
        with open(path, "r", encoding="utf-8") as f:
            text = f.read()
    except OSError:
        return None
    out: list[str] = []
    for raw in text.split("\n"):
        line = raw.strip()
        if not line or line.startswith("#"):
            continue
        out.append(line)
    return out


def run(args: argparse.Namespace, targets: list[str]) -> list[Record]:
    """Concurrency core: K worker processes, each running the full fetch+process
    pipeline. ProcessPoolExecutor.map dispatches one target at a time
    (chunksize=1) so at most K targets are in flight — the contract's bound — and
    so that variable payload sizes are load-balanced rather than batched. map
    returns results in input order, so results[i] corresponds to targets[i]."""
    n = len(targets)
    results: list[Record] = [None] * n  # type: ignore[list-item]
    if n == 0:
        return results

    k = min(args.concurrency, n)
    timeout_s = args.timeout_ms / 1000.0

    with ProcessPoolExecutor(
        max_workers=k,
        initializer=init_worker,
        initargs=(args.base_url, timeout_s),
    ) as ex:
        for i, rec in enumerate(ex.map(process_one, targets, chunksize=1)):
            results[i] = rec

    return results


def main() -> None:
    args = parse_args(sys.argv[1:])

    targets = read_targets(args.targets)
    if targets is None:
        print(f"polyfetch: cannot read targets file: {args.targets}", file=sys.stderr)
        sys.exit(EXIT_USAGE)

    try:
        records = run(args, targets)
    except Exception as e:  # catastrophic run failure (pool could not run, etc.)
        print(f"polyfetch: run failed: {e}", file=sys.stderr)
        sys.exit(EXIT_RUN_FAIL)

    write_report(records, sys.stdout)
    # Normal interpreter exit flushes stdout; exit code 0.


if __name__ == "__main__":
    main()